src/allmydata/storage/server.py

```diff
@@ -331,7 +331,31 @@ def _iter_share_files(self, storage_index):
             yield sf
 
     def remote_add_lease(self, storage_index, renew_secret, cancel_secret,
-                         owner_num=1):
+                         owner_num=1, policy=None):
+
+        if self._policy is not None:
+            if policy is None:
+                raise ValueError(
+                    "Storage Server configured with a policy, but no "
+                    "policy= kwarg was passed."
+                )
+            # XXX policy will be a dict, with one key per policy plugin
+            policy_data = policy.get(self._policy.name, None)
+            if policy_data is None:
+                raise ValueError(
+                    "Storage Server configured with a policy, but "
+                    "policy= doesn't contain '{name}'.".format(
+                        name=self._policy.name,
+                    )
+                )
+            # maybe allow_api_call should return an error-message for
+            # 'disallow' instead?
+            if not self._policy.allow_api_call(ann, "add_lease", policy_data):
+                raise RuntimeError(
+                    "Storage Server disallowed API call"
+                )
+            # okay, we've burned a token and can go ahead with the call...
+
```

Comment:
A benefit of the factoring I suggest above is that all of this code goes into a different class/method instead of being piled on top of the real add_lease logic.

Reply:
So, my instinct if we did this as "a completely separate fURL / `remote_*` method space" is to introduce another "wrapper" object that is included in Tahoe code (i.e. that has all the above code in it -- and yeah I don't want to copy-pasta that to every `remote_*` method in anycase) -- so that "plugin" authors don't have to mess with fURLs, Foolscap, etc -- that is, *ideally* Tahoe code could produce the "new `remote_*` methods for you" from your plugin + config?

(Motivation here is to make the "economic plugin" part very easy to write, even if the "tahoe part" is harder .. and so that when we switch to HTTP only the "wrapper" code has to change, NOT the actual plugin code).

I haven't thought too hard yet about whether "add one optional arg" to every method (once, now) is better or worse than "introduce a whole new fURL space for every plugin". The above is just an illustration of the former idea.

I am imagining that the `twisted/plugins/*` code -- "the plugin code" -- can live in its own separately-installed repository. So if I want the private-storage economic plugin, I can do something like `pip install privatestorage-tahoe-plugin` or so.

I feel like having the plugins *completely* wrap the API is "too much power"...? (That is, makes them complex to write and possibly ties into too many internals, etc because then you can arbitrarily change args, etc etc). That said, maybe there are use-cases where one of these things need to do that?